### ISSUE TITLE
[JIT] CodeVersionManager::LockHolder lock(Profiler ReJIT?)

### ISSUE THREAD

**Issue by hakenr** [dotnet/runtime#66649]
### Description

net6 (6.0.101) ASP.NET Core application, running in Azure App Service (Windows) with Azure Application Insights
several times a day there is a request which hangs for tens of seconds (times out)
the Application Insights profiler shows 10+ sec waiting times in `CodeVersionManager::LockHolder` when JITing dynamic code (EFCore lambdas, Newtonsoft.Json, etc.), usually in `LambdaCompiler.Compile`:
![image](https://user-images.githubusercontent.com/12828888/158353452-8f5ee45d-cb7a-4077-96fe-6ef383306f3f.png)
![image](https://user-images.githubusercontent.com/12828888/158353477-374f5578-2ab3-417d-b2d6-64b7a6b974b5.png)

It seems there is a situation where the `CodeVersionManager::LockHolder` is not able to obtain the lock and times-out.
https://github.com/dotnet/coreclr/blob/a9f3fc16483eecfc47fb79c362811d870be02249/src/vm/jitinterface.cpp#L8091-L8102

### Reproduction Steps

Unfortunately, I'm not able to provide simple repro.
It seems it is a combination of Application Insights Profiler + LambdaCompiler.Compile.

### Expected behavior

Quick LambdaCompiler.Compile run.

### Actual behavior

 LambdaCompiler.Compile() takes more than 10 sec in some cases.

### Known Workarounds

Disabling the Application Insights Profiler will probably help (will confirm later).

### Other information

You can download the .diagsessions here: https://drive.google.com/drive/folders/1mpWkhgGf5R-29Cdvf3XFnH1Aka-OCRZJ?usp=sharing

I will try to gather a memory dump from that very specific moment. Will upload later.

cc @jkotas

**Comment by jkotas** [comment]
Looks like a deadlock in ReJIT. cc @noahfalk @kouvel

**Comment by tommcdon** [comment]
@josalem

**Comment by hakenr** [comment]
Unfortunately, we were unable to take a proper memory dump to show the problem.
On the other hand, I can confirm that after disabling the Azure Application Insights Profiler, the problem completely disappeared.

**Comment by davmason** [comment]
@hakenr is disabling Application Insights a suitable workaround for you? 

I am surprised to see this deadlock, it essentially means that two separate threads are jitting methods and their inlining decisions are deadlocked on each other. Without a dump it's hard to reason about.

I wonder if another thread is involved somehow, maybe an orphaned lock? A dump would be the way to make progress

**Comment by hakenr** [comment]
@davmason Yes, disabling Application Insights (just the AI Profiler) is acceptable for us and all the issues disappeared right after that.

I'd be happy to get you the memory dump, but I haven't been successful so far and would need help with it. So far I've tried setting up an [Azure AppService Auto-Healing rule](https://azure-advice.com/2020/11/19/using-auto-heal-to-capture-memory-dumps-for-slow-requests/) that would catch the dump on a single long-running request, but it didn't create any dumps.

I can see using `procdump` directly from the AppService console as another option, but I would need help defining parameters that would catch the dump at the right time. For example, some kind of breakpoint?

**Comment by davmason** [comment]
If you can use procdump you have these options for triggers, do any of them catch the issue? Hopefully the low CPU one is applicable.

```
-C: Trigger core dump generation when CPU exceeds or equals specified value (0 to 100 * nCPU)
-c: Trigger core dump generation when CPU is less than specified value (0 to 100 * nCPU)
-M: Trigger core dump generation when memory commit limit exceeds or equals specified value (MB)
-m: Trigger core dump generation when memory commit limit is less than specified value (MB)
-T: Trigger core dump generation when thread count exceeds or equals specified value.
-F: Trigger core dump generation when filedescriptor count exceeds or equals specified value.
```

We also would need a full memory dump (-ma). Hopefully something like this would catch it.

`procdump -p <pid> -s 5 -c 50 -ma`.

That would wait for the process to have less than (50/NumCores)% CPU usage for 5 seconds before collecting a dump. I just guessed at the CPU usage, you may have to tweak it.

**Comment by hakenr** [comment]
@davmason, unfortunately, I think none of these generic conditions is applicable. It is a ASP.NET Core LOB application and the "low CPU" limit will be met most of the time.

Currently the most promising condition is "long request duration" where there will be some extra dumps but among them there should be those catching the issue. But I don't see such a condition in procdump.

Another option might a specific breakpoint, but as it is a deadlock I'm not sure we will be able to find a point which will be specific enough.

**Comment by davmason** [comment]
I can't think of any breakpoints that would catch the deadlock out of the box, but we could probably construct something.

It should be possible to create one though, I'm imagining something like your UI thread updates a timestamp every time it goes through the message loop, and a background thread checks the timestamp and triggers a breakpoint if it's been too long.

**Comment by hakenr** [comment]
Interesting idea. I should be able to create the dump directly from my code, so I just need help figuring out the conditions under which it should occur.
